CONTRIBUTING.md

```diff
@@ -5,16 +5,161 @@
+
+
+### 🆕 Extend Core Modules
+The most impactful way to contribute to LlamaIndex is extending our core modules:
+![LlamaIndex modules](docs/_static/contribution/contrib.png)
+
+
+We can roughly group the core modules into 3 categories:
+1) data ingestion & parsing, 
+2) indexing & storage, and 
+3) querying & output.
+
+We welcome contributions in _all_ modules shown above.
+So far, we have focused on implementing a core set of functionalities for each.
+As a contributor, you can help unlock the full potential.
+
+We are making rapid improvement to the project, and as a result, 
+some interfaces are still volatile.
+> Specifically, we are actively working on making the components with dotted outline more modular (core indexes, document stores, index queries, query runner). So expect volatility.
+
+#### Module Details
+Below, we will describe what each module does, give a high-level idea of the interface, show existing implementations, and give some ideas for contribution.
+
+#### Data Loaders
+A data loader ingests data of any format from anywhere into `Documents`, which can then be parsed and indexed.
+
+**Interface**: `load_data` takes arbitrary arguments as input (e.g. path to data), and outputs a sequence of `Documents`.
+
+
+**Examples**:
+* [Google Sheets Loader](https://github.com/emptycrown/llama-hub/tree/main/loader_hub/google_sheets)
+* [Gmail Loader](https://github.com/emptycrown/llama-hub/tree/main/loader_hub/gmail)
+* [Github Repository Loader](https://github.com/emptycrown/llama-hub/tree/main/loader_hub/github_repo)
+
+Contributing a data loader is easy and super impactful for the community.
+The preferred way to contribute is making a PR at [LlamaHub Github](https://github.com/emptycrown/llama-hub).
+
+**Ideas**
+* Want to load something but there's no LlamaHub data loader for it yet? Make a PR!
+
+#### Node Parser
+A node parser parses `Documents` into `Nodes` (atomic unit of data that LlamaIndex operates over, e.g., chunk of text, image, or table).
+It is responsible for splitting text (via text splitters) and explicitly modelling the relationship between units of data (e.g. A is the source of B, C is a chunk after D).
+
+**Interface**: `get_nodes_from_documents` takes a sequence of `Documents` as input, and outputs a sequence of `Nodes`.
+
+**Examples**:
+* [Simple Node Parser](https://github.com/jerryjliu/llama_index/blob/main/gpt_index/node_parser/simple.py)
+
+See [reference](https://gpt-index.readthedocs.io/en/latest/reference/node_parser.html) for full details.
+
+**Ideas**:
+* Add new `Node` relationships to model to model hierarchical document (e.g. play-act-scene, chapter-section-heading).
+
+#### Text Splitters
+Text splitter splits a long text `str` into smaller text `str` chunks with desired size and strategy. This is important since LLMs have limited context window size, and the quality of text chunk used as context impacts the quality of query results.
```

Comment:
*with a desired size and splitting "strategy"
*since LLMs have a limited context window size
*the quality of each text chunk impacts the quality of query results